From b02b9b4d93f846adc8fc8ab0a296e61ff86cc3e1 Mon Sep 17 00:00:00 2001
From: Dong Aisheng <b29396@freescale.com>
Date: Mon, 11 Nov 2013 14:53:58 +0800
Subject: [PATCH 0545/5242] MLK-11685-2 sdhci-esdhc-imx: use bus freq in
 runtime pm

commit  015728c0aca74597976a2fec6cadba20de9f7489 from
https://source.codeaurora.org/external/imx/linux-imx.git

Request BUS_FREQ_HIGH when bus is busy and then release BUS_FREQ_HIGH
when bus becomes idle.

Signed-off-by: Dong Aisheng <b29396@freescale.com>
(cherry picked from commit 64994f7115573c9ede53b51536b2c15f7cf0112a)
Signed-off-by: Haibo Chen <haibo.chen@freescale.com>

Conflicts:
	drivers/mmc/host/sdhci-esdhc-imx.c
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mmc/host/sdhci-esdhc-imx.c |   11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/drivers/mmc/host/sdhci-esdhc-imx.c b/drivers/mmc/host/sdhci-esdhc-imx.c
index c05bd17..36dff3f 100644
--- a/drivers/mmc/host/sdhci-esdhc-imx.c
+++ b/drivers/mmc/host/sdhci-esdhc-imx.c
@@ -11,6 +11,7 @@
  * the Free Software Foundation; either version 2 of the License.
  */
 
+#include <linux/busfreq-imx.h>
 #include <linux/io.h>
 #include <linux/delay.h>
 #include <linux/err.h>
@@ -1316,6 +1317,9 @@ static int sdhci_esdhc_imx_probe(struct platform_device *pdev)
 
 	pltfm_host->clk = imx_data->clk_per;
 	pltfm_host->clock = clk_get_rate(pltfm_host->clk);
+
+	request_bus_freq(BUS_FREQ_HIGH);
+
 	err = clk_prepare_enable(imx_data->clk_per);
 	if (err)
 		goto free_sdhci;
@@ -1386,6 +1390,9 @@ static int sdhci_esdhc_imx_probe(struct platform_device *pdev)
 	clk_disable_unprepare(imx_data->clk_ipg);
 disable_per_clk:
 	clk_disable_unprepare(imx_data->clk_per);
+
+	release_bus_freq(BUS_FREQ_HIGH);
+
 free_sdhci:
 	sdhci_pltfm_free(pdev);
 	return err;
@@ -1458,6 +1465,8 @@ static int sdhci_esdhc_runtime_suspend(struct device *dev)
 	}
 	clk_disable_unprepare(imx_data->clk_ahb);
 
+	release_bus_freq(BUS_FREQ_HIGH);
+
 	return ret;
 }
 
@@ -1468,6 +1477,8 @@ static int sdhci_esdhc_runtime_resume(struct device *dev)
 	struct pltfm_imx_data *imx_data = sdhci_pltfm_priv(pltfm_host);
 	int err;
 
+	request_bus_freq(BUS_FREQ_HIGH);
+
 	err = clk_prepare_enable(imx_data->clk_ahb);
 	if (err)
 		return err;
-- 
1.7.9.5

